src/dmd/traits.d

```diff
@@ -1580,7 +1580,7 @@ Expression semanticTraits(TraitsExp e, Scope* sc)
                     //printf("p.loc.linnum = %d\n", p.loc.linnum);
 
                     o = p.parseTypeOrAssignExp(TOK.endOfFile);
-                    if (errors != global.errors || p.token.value != TOK.endOfFile)
+                    if (p.token.value != TOK.endOfFile)
```

Comment:
Although it seems redundant to check for a parse error here, it should do no harm to skip semantic analysis if a parse error has occurred. Not running the analysis on a lot of failing `__traits(compiles)` might save compile time.

I don't the problem is here. IIRC template analysis bails out unconditionally after any other error.

Reply:
`errors == global.gaggedErrors`.  So the check is always true!

Reply:
If you really want to keep the check, then it's going to have to be:
```
const errors2 = global.errors;
if (errors2 != global.errors || p.token.value != TOK.endOfFile)
   ...
```

Reply:
> errors == global.gaggedErrors

Not sure what you mean here.


> If you really want to keep the check, then it's going to have to be:

Do you mean that there should be a local error variable for each iteration? I don't think that is necessary as any error breaks the loop.

Reply:
Ah, sorry, just noticed that startGagging doesn't return errors, but gaggedErrors. I'll have to reconsider...